infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -607,35 +685,35 @@ public void onClusterRoleChange(TopologyDescriptor newTopology) {
         // We do not update topology until we successfully stop log replication
         if (localClusterDescriptor.getRole() == ClusterRole.ACTIVE) {
             stopLogReplication();
+            logReplicationEventListener.stop();
         }
 
-        // Update topology, cluster, and node configs
-        updateLocalTopology(newTopology);
-
-        // Update topology config id in metadata manager
-        logReplicationMetadataManager.setupTopologyConfigId(topologyDescriptor.getTopologyConfigId());
+        // Stop the replication server
+        if (localClusterDescriptor.getRole() == ClusterRole.STANDBY) {
+            interClusterServerNode.disable();
+            logReplicationServerFuture.cancel(true);
+        }
 
         if (isLeader.get()) {
-            // Reset the Replication Status on Active and Standby only for the leader node
-            // Consider the case of async configuration changes, non-lead nodes could overwrite
-            // the replication status if it has already completed by the lead node
+            // Reset the Replication Status on Active and Standby only on the
+            // leader node.  Consider the case of async configuration changes,
+            // non-lead nodes could overwrite the replication status if it
+            // has already completed by the lead node
             resetReplicationStatusTableWithRetry();
         }
 
-        log.debug("Persist new topologyConfigId {}, cluster id={}, role={}", topologyDescriptor.getTopologyConfigId(),
-                localClusterDescriptor.getClusterId(), localClusterDescriptor.getRole());
+        // Clear existing Sink and Metadata Managers.
+        clientToMetadataManagerMap.clear();
+        clientToSinkManagerMap.clear();
 
-        // Update replication manager
-        updateReplicationManagerTopology(newTopology);
+        // Update topology, cluster, and node configs
+        updateLocalTopology(newTopology);
```

Comment:
actually, we do not need to update the topology in this case because:

1. If the new role is Source, a new replicationManager with the new topology will get created.
2. If the new role is Sink, the replicationManager is not relevant.